src/interpreter/executors/main.rs

```diff
@@ -264,25 +264,27 @@ where
         tx: CheckedTransaction,
         params: ConsensusParameters,
     ) -> Result<StateTransition, InterpreterError> {
-        Interpreter::with_storage(storage, params)
+        let mut interpreter = Interpreter::with_storage(storage, params);
+        interpreter
             .transact(tx)
-            .map(|st| st.into_owned())
+            .map(|state| StateTransition::new(state, interpreter.tx.into(), interpreter.receipts))
     }
 
     /// Initialize a pre-allocated instance of [`Interpreter`] with the provided
     /// transaction and execute it. The result will be bound to the lifetime
     /// of the interpreter and will avoid unnecessary copy with the data
     /// that can be referenced from the interpreter instance itself.
-    pub fn transact(&mut self, tx: CheckedTransaction) -> Result<StateTransitionRef<'_>, InterpreterError> {
+    pub fn transact(&mut self, tx: CheckedTransaction) -> Result<ProgramState, InterpreterError> {
         let state_result = self.init_script(tx).and_then(|_| self.run());
 
         #[cfg(feature = "profile-any")]
         self.profiler.on_transaction(&state_result);
 
-        let state = state_result?;
-
-        let transition = StateTransitionRef::new(state, self.transaction(), self.receipts());
+        state_result
+    }
 
-        Ok(transition)
+    /// Get the state transaction from a new program state.
+    pub fn state_transition<'a>(&'a self, state: &'a ProgramState) -> StateTransitionRef<'a> {
```

Comment:
Now the API looks not consistent=( First, you need to call `transact`, and after with the output, you need to call `state_transition`.

It seems that `StateTransitionRef` still can own `ProgramState` as before, but `Transactor` should work with `ProgramState`.

In this case, API will be consistent, and we will remove `unsafe`=)
But you need to add a separate method like `transact_internal` that returns `ProgramState`. 

Also maybe `StateTransitionRef` can be generic like `StateTransitionRef<'a, Generic>` where in some cases it is `&'a PrgoramState` in some `ProgramState`(it will remove unnecessary `clone` in some cases).

Reply:
I don't think that's possible though because the `ProgramState` will be created in the `transact_internal` and we can't return a reference to `ProgramState` when it's a local variable in the function.

Reply:
`StateTransitionRef<'a, ProgramState>` will own `ProgramState` inside=) `transact_internal` it is for `StateTransaciton`

Reply:
@xgreenx I just pushed some changes, does that fix your concern?